### PR TITLE
Prevent #links to work except in community chats and allow redirection from channel to another when clicked

### LINKED ISSUE #15048: Handle #links in chats
#links in chat if clicked will open a public chat.
This is not necessary anymore as we won't support them in the future.

Clicking on #links should be disabled on 1-to-1 chats and private group chats. They should not appear highlighted if possible.

Clicking on #links on a community should take you to the channel with the same name, if a channel with that name does not exist, nothing should happen.

There's multiple ways to implement this, please ping me before implementing it.

### PR DESCRIPTION
**Pull request by ibrkhalil**
fixes #15048
status: needs QA review

### CODE REVIEW

**Review by cammellos — APPROVED** [review]

File: src/status_im2/subs/chat/chats.cljs
Comment (cammellos) [inline src/status_im2/subs/chat/chats.cljs]: You can also re-use `chats/chat` and extract the `community-id` in this sub, something like this should work I believe:

```
(re-frame/reg-sub
  :community-id-by-chat-id
  (fn [[_ chat-id]]
    [(re-frame/subscribe [:chats/chat chat-id])])
  (fn [[chat]]
    (:community-id chat)))
```

**Comment by pavloburykh** [comment]
Hey @ibrkhalil Thanx for the PR! Could you please rebase it to the latest develop?

**Comment by pavloburykh** [comment]
@ibrkhalil thanx for the PR. Please, take a look at the following issues

### ISSUE 1 Hashtag links are not resolved in communities

**Preconditions:** create a community (for example "test community") with at least 2 channels in it (for example #general and #test channel )

**Steps:**

1. Open "test community"
2. Open #test channel
3. Send message with hashtag link #general
4. See if link is resolved 

**Expected result:** link is resolved. Clicking on link redirects user to the #general channel 

**Actual result:** link is not resolved. Unable to click the link and redirect to #general channel 


https://user-images.githubusercontent.com/97245802/220597259-28d4218d-657a-434e-afef-8420fcc6e366.mp4

**Comment by pavloburykh** [comment]
### ISSUE 2 Hashtags are removed in 1-1 and private group chats 

**Steps:**

1. Open 1-1 or group chat
2. Send message with word containing # at the beginning (for example #message) 
3. See if hashtag remains in sent message

**Expected result:** hashtags should not be removed. Such message just shouldn't be resolved as link (highlighted) but as a regular text. 

**Actual result:** hashtag is automatically removed. At the same time it is visible in chat preview. 

https://user-images.githubusercontent.com/97245802/220599138-1963c07f-58a2-4657-a20f-b634f8c695c9.mp4

File: src/status_im/ui2/screens/chat/messages/message.cljs
Comment (ilmotta) [inline src/status_im/ui2/screens/chat/messages/message.cljs:99]: Just a side note, not that you need to change anything.

We could remove the type hint on the `literal` argument. It's really odd to add a type hint to a string in Clojure, because Clojure already provides lots of functions to process strings, so why would we use the Javascript ones? 

```clojure
;; This is unnecessary.
(.substring literal 0 (.-length literal))

;; Good
(subs literal 0 (count literal))
```

**Comment by ibrkhalil** [comment]
> @ibrkhalil thanx for the PR. Please, take a look at the following issues
> 
> ### ISSUE 1 Hashtag links are not resolved in communities
> **Preconditions:** create a community (for example "test community") with at least 2 channels in it (for example #general and #test channel )
> 
> **Steps:**
> 
> 1. Open "test community"
> 2. Open #test channel
> 3. Send message with hashtag link #general
> 4. See if link is resolved
> 
> **Expected result:** link is resolved. Clicking on link redirects user to the #general channel
> 
> **Actual result:** link is not resolved. Unable to click the link and redirect to #general channel

There was a PR that was merged that removed this functionality, I added it again

**Comment by ibrkhalil** [comment]
> ### ISSUE 2 Hashtags are removed in 1-1 and private group chats
> **Steps:**
> 
> 1. Open 1-1 or group chat
> 2. Send message with word containing # at the beginning (for example #message)
> 3. See if hashtag remains in sent message
> 
> **Expected result:** hashtags should not be removed. Such message just shouldn't be resolved as link (highlighted) but as a regular text.
> 
> **Actual result:** hashtag is automatically removed. At the same time it is visible in chat preview.

Fixed in last commit

File: src/status_im/ui2/screens/chat/messages/message.cljs
Comment (ibrkhalil) [reply src/status_im/ui2/screens/chat/messages/message.cljs:99]: Thanks Icaro for the input, I have removed it

File: src/status_im2/subs/chat/chats.cljs
Comment (ibrkhalil) [reply src/status_im2/subs/chat/chats.cljs]: Used it, Thanks Andrea!

**Comment by pavloburykh** [comment]
@ibrkhalil thanx for fixing issues 1 and 2

I have one more question regarding validation rules for channel names in hashtag links.
Currently we have different validation rules for channel names on desktop and mobile. 
- On mobile all symbols are allowed, i.e. emoji, special characters (e.g. " , / ! ^ . etc.)
- On desktop only latin letters, numbers, underscores, hypthens are allowed. Whitespaces are automatically replaces by hypthens in channel names on desktop. 

Based on what I currently observe,  in hashtag links we support only those validation rules which are applied on desktop. It means that if user creates channel on mobile  with name "#test!@_channel name" it will be not resolved as hashtag link. 
@ibrkhalil please verify that such behaviour is correct. I suppose that this is correct, cause desktop channel name rules are prioritised. I am not sure if we are going to have community management feature on mobile in future, but if we are - then we need to synchronise validation rules for channel names on mobile and desktop. Anyway, this is not in scope of current PR. 

Still even if we are following desktop rules there is an issue with hashtag links to the channels which contain underscores in its name (see details below)

### ISSUE 3 Hashtag links to the channels which contain text with underscores around are not resolved as links

**Steps:** 

1. In community create a channel that has text with underscores around it. For example: \_test-channel\_
2. In another channel of current community post a hashtag link to the newly created channel #\_test-channel\_ 
**Note:** do not forget to use `\`symbol  before underscore `_` characters in order to ignore Markdown formatting, otherwise you will see italic text in hashtag link. 
4. See if hashtag link is resolved. 

**Expected result:** hashtag link is resolved. User is redirected to #\_test-channel\_ channel on clicking the link 

**Actual result:** hashtag link is not resolved. Regular text is sent. 


https://user-images.githubusercontent.com/97245802/221606006-4b37f2a7-716c-42b9-93c2-04366730c12a.mp4

**Comment by ibrkhalil** [comment]
> @ibrkhalil thanx for fixing issues 1 and 2
> 
> I have one more question regarding validation rules for channel names in hashtag links. Currently we have different validation rules for channel names on desktop and mobile.
> 
> * On mobile all symbols are allowed, i.e. emoji, special characters (e.g. " , / ! ^ . etc.)
> * On desktop only latin letters, numbers, underscores, hypthens are allowed. Whitespaces are automatically replaces by hypthens in channel names on desktop.
> 
> Based on what I currently observe, in hashtag links we support only those validation rules which are applied on desktop. It means that if user creates channel on mobile with name "#test!@_channel name" it will be not resolved as hashtag link. @ibrkhalil please verify that such behaviour is correct. I suppose that this is correct, cause desktop channel name rules are prioritised. I am not sure if we are going to have community management feature on mobile in future, but if we are - then we need to synchronise validation rules for channel names on mobile and desktop. Anyway, this is not in scope of current PR.
> 
> Still even if we are following desktop rules there is an issue with hashtag links to the channels which contain underscores in its name (see details below)
> 
> ### ISSUE 3 Hashtag links to the channels which contain text with underscores around are not resolved as links
> **Steps:**
> 
> 1. In community create a channel that has text with underscores around it. For example: _test-channel_
> 2. In another channel of current community post a hashtag link to the newly created channel #_test-channel_
>    **Note:** do not forget to use `\`symbol  before underscore `_` characters in order to ignore Markdown formatting, otherwise you will see italic text in hashtag link.
> 3. See if hashtag link is resolved.
> 
> **Expected result:** hashtag link is resolved. User is redirected to #_test-channel_ channel on clicking the link
> 
> **Actual result:** hashtag link is not resolved. Regular text is sent.

Will investigate and fix after the current issue I am working on

**Comment by ibrkhalil** [comment]
> > @ibrkhalil thanx for fixing issues 1 and 2
> > I have one more question regarding validation rules for channel names in hashtag links. Currently we have different validation rules for channel names on desktop and mobile.
> > 
> > * On mobile all symbols are allowed, i.e. emoji, special characters (e.g. " , / ! ^ . etc.)
> > * On desktop only latin letters, numbers, underscores, hypthens are allowed. Whitespaces are automatically replaces by hypthens in channel names on desktop.
> > 
> > Based on what I currently observe, in hashtag links we support only those validation rules which are applied on desktop. It means that if user creates channel on mobile with name "#test!@_channel name" it will be not resolved as hashtag link. @ibrkhalil please verify that such behaviour is correct. I suppose that this is correct, cause desktop channel name rules are prioritised. I am not sure if we are going to have community management feature on mobile in future, but if we are - then we need to synchronise validation rules for channel names on mobile and desktop. Anyway, this is not in scope of current PR.
> > Still even if we are following desktop rules there is an issue with hashtag links to the channels which contain underscores in its name (see details below)
> > ### ISSUE 3 Hashtag links to the channels which contain text with underscores around are not resolved as links
> > **Steps:**
> > 
> > 1. In community create a channel that has text with underscores around it. For example: _test-channel_
> > 2. In another channel of current community post a hashtag link to the newly created channel #_test-channel_
> >    **Note:** do not forget to use `\`symbol  before underscore `_` characters in order to ignore Markdown formatting, otherwise you will see italic text in hashtag link.
> > 3. See if hashtag link is resolved.
> > 
> > **Expected result:** hashtag link is resolved. User is redirected to #_test-channel_ channel on clicking the link
> > **Actual result:** hashtag link is not resolved. Regular text is sent.
> 
> Will investigate and fix after the current issue I am working on

@cammellos Can I have access to our markdown repo to allow for underscores in `status-tags` parsing ?
I will do the validation side of status-mobile when I get a response from the design team on how it's supposed to look when a user enters a bad name.

**Comment by ibrkhalil** [comment]
> @ibrkhalil thanx for fixing issues 1 and 2
> 
> I have one more question regarding validation rules for channel names in hashtag links. Currently we have different validation rules for channel names on desktop and mobile.
> 
> * On mobile all symbols are allowed, i.e. emoji, special characters (e.g. " , / ! ^ . etc.)
> * On desktop only latin letters, numbers, underscores, hypthens are allowed. Whitespaces are automatically replaces by hypthens in channel names on desktop.
> 
> Based on what I currently observe, in hashtag links we support only those validation rules which are applied on desktop. It means that if user creates channel on mobile with name "#test!@_channel name" it will be not resolved as hashtag link. @ibrkhalil please verify that such behaviour is correct. I suppose that this is correct, cause desktop channel name rules are prioritised. I am not sure if we are going to have community management feature on mobile in future, but if we are - then we need to synchronise validation rules for channel names on mobile and desktop. Anyway, this is not in scope of current PR.
> 
> Still even if we are following desktop rules there is an issue with hashtag links to the channels which contain underscores in its name (see details below)
> 
> ### ISSUE 3 Hashtag links to the channels which contain text with underscores around are not resolved as links
> **Steps:**
> 
> 1. In community create a channel that has text with underscores around it. For example: _test-channel_
> 2. In another channel of current community post a hashtag link to the newly created channel #_test-channel_
>    **Note:** do not forget to use `\`symbol  before underscore `_` characters in order to ignore Markdown formatting, otherwise you will see italic text in hashtag link.
> 3. See if hashtag link is resolved.
> 
> **Expected result:** hashtag link is resolved. User is redirected to #_test-channel_ channel on clicking the link
> 
> **Actual result:** hashtag link is not resolved. Regular text is sent.

Hey Pavlo, After chatting with Andrea on this he told me it's better to do this in a separate issue .. Can we do that and also unblock this PR ?

**Comment by pavloburykh** [comment]
> Hey Pavlo, After chatting with Andrea on this he told me it's better to do this in a separate issue .. Can we do that and also unblock this PR ?

@ibrkhalil totally agree, I wanted to propose the same as this issue is an edge case and not a blocker. 

Let me perform another smoke to make sure no regression appeared after recent rebasing and re-run e2e. After that we will be ready to merge the PR. Will ping you as soon as I finish. Thanx!

**Comment by pavloburykh** [comment]
@ibrkhalil E2E have found a regression, please take a look

### ISSUE 4 Edited mark is not displayed for edited messages

**Steps:** 

1. Open any chat
2. Send a message
3. Edit message 
4. See if edited mark is displayed near edited message

**Expected result:** edited mark is displayed

**Actual result:** edited mark is not displayed


https://user-images.githubusercontent.com/97245802/224734013-8fe5aefa-641d-4da8-9673-bcc1e56cea33.mp4

**Comment by ibrkhalil** [comment]
> @ibrkhalil E2E have found a regression, please take a look
> 
> ### ISSUE 2 Edited mark is not displayed for edited messages
> **Steps:**
> 
> 1. Open any chat
> 2. Send a message
> 3. Edit message
> 4. See if edited mark is displayed near edited message
> 
> **Expected result:** edited mark is displayed
> 
> **Actual result:** edited mark is not displayed

Sorry for that, Fixed

**Comment by pavloburykh** [comment]
@ibrkhalil thanx for the fix. PR is ready to be merged. I create a separate ticket for ISSUE 3 after this PR will be merged.

File: src/status_im/communities/core.cljs
Comment (flexsurfer) [inline src/status_im/communities/core.cljs:213]: `:dispatch` should be used only as an exception

File: src/status_im/ui2/screens/chat/messages/message.cljs
Comment (flexsurfer) [inline src/status_im/ui2/screens/chat/messages/message.cljs:85]: colors must be from colors ns

File: src/status_im2/contexts/chat/messages/content/text/view.cljs
Comment (flexsurfer) [inline src/status_im2/contexts/chat/messages/content/text/view.cljs:67]: colors must be from colors ns

File: src/status_im/communities/core.cljs
Comment (flexsurfer) [reply src/status_im/communities/core.cljs:213]: `navigate-to-chat` event should be used directly

File: src/status_im/ui2/screens/chat/messages/message.cljs
Comment (ibrkhalil) [reply src/status_im/ui2/screens/chat/messages/message.cljs:85]: Will make a follow up tomorrow

File: src/status_im2/contexts/chat/messages/content/text/view.cljs
Comment (ibrkhalil) [reply src/status_im2/contexts/chat/messages/content/text/view.cljs:67]: Will make a follow up tomorrow

File: src/status_im/communities/core.cljs
Comment (ibrkhalil) [reply src/status_im/communities/core.cljs:213]: Will make a follow up tomorrow

**Comment by du82** [comment]
This doesn't work for me, it just removes the hashtag:

https://user-images.githubusercontent.com/65834069/224881612-81a55774-1679-4d77-a5a9-3440d11a5201.mp4

**Comment by flexsurfer** [comment]
@ibrkhalil see comment above

**Pull request merged**
